types/query/filtered_pagination.go

```diff
@@ -114,3 +115,122 @@ func FilteredPaginate(
 
 	return res, nil
 }
+
+func GenericFilteredPaginate[V any, F codec.ProtoMarshaler](
+	cdc codec.BinaryCodec,
+	prefixStore types.KVStore,
+	pageRequest *PageRequest,
+	onResult func(key []byte, value V) (F, error),
+) ([]F, *PageResponse, error) {
+	// if the PageRequest is nil, use default PageRequest
+	if pageRequest == nil {
+		pageRequest = &PageRequest{}
+	}
+
+	offset := pageRequest.Offset
+	key := pageRequest.Key
+	limit := pageRequest.Limit
+	countTotal := pageRequest.CountTotal
+	reverse := pageRequest.Reverse
+	results := []F{}
+
+	if offset > 0 && key != nil {
+		return results, nil, fmt.Errorf("invalid request, either offset or key is expected, got both")
+	}
+
+	if limit == 0 {
+		limit = DefaultLimit
+
+		// count total results when the limit is zero/not supplied
+		countTotal = true
+	}
+
+	if len(key) != 0 {
+		iterator := getIterator(prefixStore, key, reverse)
+		defer iterator.Close()
+
+		var numHits uint64
+		var nextKey []byte
+
+		for ; iterator.Valid(); iterator.Next() {
+			if numHits == limit {
+				nextKey = iterator.Key()
+				break
+			}
+
+			if iterator.Error() != nil {
+				return nil, nil, iterator.Error()
+			}
+
+			protoMsg := any(new(V))
+			err := cdc.Unmarshal(iterator.Value(), protoMsg.(codec.ProtoMarshaler))
```

Comment:
so `V` is an `any`, which is just an `interface{}`. Do we not need to type check here that `V` is a `codec.ProtoMarshaler`? Why not also make `V` of type `codec.ProtoMarshaler`?

Reply:
I was trying to find a way of instantiating `V` (now `T`; not sure what to name these), which for some reason was impossible for me (always ended up getting `**authz.Grant` instead of `*authz.Grant` without the possibility to dereference it). With `reflect`it was quite easy, but I'm not sure if we would like to use it here. See my last commit: https://github.com/cosmos/cosmos-sdk/pull/12253/commits/6db3de3d85fae502b7334f14869185564e7b1031

Reply:
This was what I was trying to do before:

```
protoMsg := any(new(T)).(codec.ProtoMarshaler)
protoMsg := any(*new(T)).(codec.ProtoMarshaler)
```

None of these work... In short, we need to be able to initialize a variable with type `T` that is not nil. Is this possible with generics?

Reply:
In Osmosis, I made V of type proto.Message for something similar here; https://github.com/osmosis-labs/osmosis/blob/main/tests/e2e/chain/config.go#L149-L162

Reply:
`codec.BinaryCodec` calls for `codec.ProtoMarshaler`. It shouldn't be an issue, it seems like all of our generated protobuf structs are `codec.ProtoMarshaler`.
The issue we were discussing here has been solved by adding a constructor function.